grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy

```diff
@@ -338,77 +336,30 @@ class StockMovementApiController {
-                if (lotNumber?.contains("E") && NumberUtils.isNumber(lotNumber)) {
-                    throw new IllegalArgumentException("Lot numbers must not be specified in scientific notation. " +
-                            "Please reformat field with Lot Number: \"${lotNumber}\" to a number format")
-                }
-
-                PickPageItem pickPageItem = pickPageItems.find {
-                    it.requisitionItem?.id == requisitionItemId
-                }
-
-                if (!pickPageItem) {
-                    throw new IllegalArgumentException("Requisition item id: ${requisitionItemId} not found")
-                }
-
-                // FIXME Should find bin location by name and parent and inventory item by lot number and expiration date
-                // and compare object equality (or at least PK equality) rather than comparing various components
-                AvailableItem availableItem = pickPageItem.availableItems?.find {
-                    (binLocation ? it.binLocation?.name == binLocation : !it.binLocation) && lotNumber == (it.inventoryItem?.lotNumber ?: null) &&
-                            expirationDate == (it?.inventoryItem?.expirationDate ? it.inventoryItem.expirationDate.format(Constants.EXPIRATION_DATE_FORMAT) : null)
-                }
+        MultipartFile importFile = command.importFile
+        if (importFile.isEmpty()) {
+            throw new IllegalArgumentException("File cannot be empty")
+        }
 
-                if (!availableItem) {
-                    throw new IllegalArgumentException("There is no item available with lot: ${lotNumber ?: ""}, expiration date: ${tokens[2] ?: ""} and bin: ${binLocation ?: ""}")
-                }
+        String csv = new String(importFile.bytes)
+        List<Map> importedLines = stockMovementService.parsePickCsvTemplateImport(csv)
 
-                RequisitionItem requisitionItem = pickPageItem.requisitionItem?.modificationItem ?: pickPageItem.requisitionItem
+        stockMovementService.validatePicklistListImport(importedLines, pickPageItems)
 
-                pickPageItem.picklistItems.each {
-                    if (it.id) {
-                        it.quantity = 0
-                    }
-                }
-                pickPageItem.picklistItems.add(new PicklistItem(
-                        requisitionItem: requisitionItem,
-                        inventoryItem: availableItem.inventoryItem,
-                        binLocation: availableItem.binLocation,
-                        quantity: quantityPicked,
-                        sortOrder: pickPageItem.sortOrder
-                ))
-            }
+        List errors = importedLines*.errors.withIndex().collect { errors, index ->
+            errors.collect { "Row ${1}: ${it}" }
+        }.flatten()
 
-            stockMovementService.createOrUpdatePicklistItem(stockMovement, pickPageItems)
+        stockMovementService.importPicklistTemplate(importedLines, stockMovement, pickPageItems)
```

Comment:
why we first import the template and then check if the errors are not empty?

Reply:
some of the requirements for import are

 _if I enter a stock record that does not exist in inventory, then the pick isn’t imported for these incorrect lines, and all valid lines show the pick from the import successfully and the invalid lines remain empty and are highlighted in red to complete the pick_

which means that user expect to import valid lines and get errors for invalid ones.

The way Manon explained this to me is that user might import the file, export pick file, fix rows that were invalid and were not imported and then import the file again.

Basically main point is that we don't want to fail the import if there are invalid lines.
If you guys disagree with that, then this needs to be discussed with Manon and Kelsey.

Reply:
Maybe from a user perspective it does have a sense, but from a best practices/REST/anything perspective, this smells to me.

Reply:
Moreover, prove me wrong, but in case of situation when a few items have errors and user would want to fix them, I assume the user wouldn't remove the valid ones (those that are already persisted) from the sheet.
Would we actually update an existing item or we would create a new pick list item?
 Looking quickly at the code, I'm concerned that we actually wouldn't update the existing item.

Reply:
Sorry, for some reason I'm not seeing all comments when I click on Files Changed tab so I missed this one the first time through. 

> Maybe from a user perspective it does have a sense, but from a best practices/REST/anything perspective, this smells to me.

I know we've talked about this a bit already but just wanted to provide some context in case someone stumbles upon this conversation and gets confused. 

Remember, this isn't supposed to be a REST API. This is a multi-part request that was implemented as a REST endpoint. Not to say it couldn't be implemented as a REST endpoint (see the proposals discussed earlier
https://pihemr.atlassian.net/wiki/spaces/OB/pages/3075342378/Data+Imports), but we're simply uploading a file and processing the file on the server as we would an XLS or CSV file. Imagine this happening from an HTML form. 

The difference between what we're doing here and uploading a form via a GSP is that we usually don't have React  interacting with the server via standard HTTP requests. And perhaps we shouldn't ... but until we create a convention around importing data via the API, it's probably safe to go with the HTTP approach. 
https://pihemr.atlassian.net/wiki/spaces/OB/pages/3075342378/Data+Imports

So our convention with uploaded data files (i.e. Configuration > Import Data) has been to prevent data from being saved unless all the data is valid. But the case that Manon wants to implement here is slightly different from most of those because the picklist item data we're importing are essentially "edits" to the existing picklist. That's not entirely true since the user will be importing brand new picklist items after clearing the picklist. 

The confusion here might be due to the fact that we're exposing this endpoint to a React client where the expectation is to handle a multi-part request and return a JSON response i.e. we have to return the errors as JSON. I don't know if there's a better way for the server to deal with this type of interaction. But since Axios is an HTTP client (not a REST) client it seems reasonable to implement it this way. 

Last point. You're correct that this might be handled differently if it was a REST endpoint. But there's precedence for implementing this type of "save what's valid, return errors for what's not". I have a bunch of resources saved that describe this "bulk api" behavior. 

Here are a few:

* https://tyk.io/blog/api-design-guidance-bulk-and-batch-import/
* https://stackoverflow.com/questions/28596688/rest-api-bulk-create-or-update-in-single-request

> Moreover, prove me wrong, but in case of situation when a few items have errors and user would want to fix them, I assume the user wouldn't remove the valid ones (those that are already persisted) from the sheet.

This is a good point. But my understanding is that uploading the file multiple times would not create duplicate picklist items (which I think is your concern). It would just add new lines and update lines that have changed. Furthermore, removing lines from the import file might actually signify that we want to remove them. But that is not a requirement here, as far as I know, so just ignore that point.

In either case, this would be a good test case to implement in an integration test. 

* User clears picklist 
* User uploads a data file with 10 lines, 8 valid, 2 invalid
* System creates 8 picklist items
* User fixes 2 invalid lines
* User uploads a data file with 10 lines
* System ignores 8 valid lines, creates 2 picklist items

We could even add in a listener on Hibernate events to make sure we're not touching the 8 valid lines on the second import.